portfolio/src/main/java/com/google/sps/servlets/ChartServlet.java

```diff
@@ -0,0 +1,46 @@
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.lang.String;
+
+
+@WebServlet("/chart-data")
+public class ChartServlet extends HttpServlet {
+
+  private Map<String, Integer> popularCommenters = new HashMap<>();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    Gson gson = new Gson();
+    String json = gson.toJson(popularCommenters);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {  
+    if (("null").equals(request.getParameter("deleted"))) {
```

Comment:
We can reduce the code complexity by reducing nest if-else. Usually we can consider if the nest if-else can be simplified when there are more than 2 levels deep. You code looks OK. But let's compare it with another way using early return.
```java
if (("null").equals(request.getParameter("deleted"))) {
    popularCommenters.clear();
    return;
}

String email = request.getParameter("email");
if (email == null) {
    return;
}

int currentVotes = popularCommenters.containsKey(email) ? popularCommenters.get(email) : 0;
Boolean deleted = Boolean.parseBoolean(request.getParameter("deleted"));
if (!deleted) {
    popularCommenters.put(email, currentVotes + 1);
} else {
    // This is the case where the comment is deleted not submitted
    if (currentVotes != 0) {
        popularCommenters.put(email, currentVotes - 1);
    }
}

```

Reply:
I think this version is also better in that it is clearer to read. I tried yesterday doing what is below, after the Java 101 session to eliminate the outermost nested if condition so I tried doing what's below but when I tried parsing the null to a Boolean it was always skipping the first if condition and say that deleted is false and I don't understand why? I still think your version is better and clearer, but I am just curious why mine wasn't working.

` @Override
  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {  
    String email = request.getParameter("email");
    Boolean deleted = Boolean.parseBoolean(request.getParameter("deleted"));
    if (deleted == null) {
      popularCommenters.clear();
    } else if (email != null && !deleted) {
      int currentVotes = popularCommenters.containsKey(email) ? popularCommenters.get(email) : 0;
      popularCommenters.put(email, currentVotes + 1);
    } else if (email != null && deleted) { // This is the case where the comment is deleted not submitted
      int currentVotes = popularCommenters.containsKey(email) ? popularCommenters.get(email) : 0;
      if (currentVotes != 0) {
        popularCommenters.put(email, currentVotes - 1);
      }
    }
  }`

Reply:
Also, I am sorry for that horrible formatting, I tried using the code button but I dont know why its coming out like that.